cpp/src/lance/arrow/dataset.cc

```diff
@@ -382,4 +391,52 @@ ::arrow::Result<::arrow::dataset::FragmentIterator> LanceDataset::GetFragmentsIm
   return ::arrow::MakeVectorIterator(fragments);
 }
 
+::arrow::Result<std::shared_ptr<LanceDataset>> LanceDataset::AddColumns(
+    const std::shared_ptr<::arrow::Table>& other,
```

Comment:
Would it make sense for the interface to be Dataset instead of Table? At the python level we can convert Table to In-memory Dataset or something? And it would also support FilesystemDataset? Because if one table is larger than memory, it might not be uncommon for the other to also be so? Again, if it's not a big refactor, we could also make this change later if that makes more sense

Reply:
It could raise other issues, i.e., now we assume that table is in memory, so that we can do in-memory hash join. 
If it is a dataset, it becomes unbound to memory, which is more complicated, i.e., might need external sort and etc.